### PR TITLE
Beginning of doc guide

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Several committers have broached the idea of having a guide for how we document Ruby.  Here's a proposal for a couple of ideas.  There will need to be much more than just Synopsis in the section on Structure, so this is truly just the beginning.  I'll be adding more as we go, to say nothing of modifying per committers' comments.

### CODE REVIEW

File: doc/guides/method.rdoc [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+When a method has multiple signatures,
+document each signature separately, as if it were a distinct method.
+
+In <tt>array.c</tt>, for example, we have this:
+  *  call-seq:
+  *    array[index] -> object or nil
+  *    array[start, length] -> object or nil
+  *    array[range] -> object or nil
+  *    array.slice(index) -> object or nil
```
Comment (jeremyevans) [inline doc/guides/method.rdoc:25]: I think we should not list all aliases as separate lines in `call-seq`.  We should pick one as the canonical method, and have the other one be documented as `alias for $canonical_method`.

File: doc/guides/method.rdoc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  *    array.slice(start, length) -> object or nil
+  *    array.slice(range) -> object or nil
+
+Document each signature separately, beginning with a level-6 header:
```
Comment (jeremyevans) [inline doc/guides/method.rdoc:29]: I prefer to only use headings in method documentation in exceptional circumstances, but we definitely shouldn't jump headings.  You should only use a level-6 heading if you are currently under a level-5 heading.

File: doc/guides/method.rdoc [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  *    array <=> other_array -> -1, 0, or 1
+  *
+  *  Returns -1, 0, or 1 as +self+ is less than, equal to, or greater than +other_array+.
+
+== Multiple Signatures
+
+\[Is 'signatures' the correct term?]
+
+When a method has multiple signatures,
+document each signature separately, as if it were a distinct method.
```
Comment (jeremyevans) [inline doc/guides/method.rdoc:18]: I'm not sure this is a good general rule, though I think it does make sense in this case.  Whether the signatures should be documented separately depends on how much overlap there is between them.

**Comment by BurdetteLamar** [comment]
Not sure I want to take this on just now.

**Comment by ioquatix** [comment]
@jeremyevans since you suggested this, maybe you can bring this PR to completion? Because we definitely want @BurdetteLamar to feel confident in the value of their contributions, which so far have been epic. Even without the consistency prescribed by this proposed guide, the documentation has been made so much better by a huge effort of @BurdetteLamar, and I while critical discussion should always be welcome, I think we also need to respect the huge effort that goes into writing better documentation.

@BurdetteLamar thanks for your amazing work so far, it has been a huge improvement in every possible way.

**Comment by marcandre** [comment]
I agree with @jeremyevans on all his points.

I agree that @BurdetteLamar's contributions have been great I hope that this continues. I also hope that the criticism will be welcomed and addressed with much less resistance.

**Comment by ioquatix** [comment]
> addressed with much less resistance.

In the past year, @BurdetteLamar has merged ~50 pull requests comprising significant improvements to the Ruby documentation.

Looking at some recent PRs I can see that the discussion has been somewhat confrontational.

Ruby is an open source project supported by a community of people working for free.

There should be no expectation of anyone doing anything. That includes: 1/ making PRs, 2/ giving feedback, 3/ addressing feedback and 4/ merging commits.

My understanding is that @BurdetteLamar is doing this work to improve Ruby primarily for enjoyment. I can certainly understand that if the communication becomes confrontational, any enjoyment is quickly lost.

I completely understand the need for the documentation to be correct. But short of an existing guide (i.e. this PR), I don't think we can criticise anyone for looking at the existing documentation and trying to build a consistent model out of it and running with it. Frankly speaking, the documentation wasn't that great to start with, so it's no surprise that there are structural issues.

I don't want to see people get frustrated with each other because they care about Ruby being the best it can be. So, can we all just take a breather? We are all in this together and we all have the same goal.

**Comment by jeremyevans** [comment]
I don't want my tone to be confrontational, and I apologize if it was.  In general, I only reviewed the pull requests that @BurdetteLamar asked me to, and provided candid feedback with the goal of making the documentation more useful for all Ruby users.  I definitely appreciate @BurdetteLamar's contributions and hope they will continue.  I didn't want @BurdetteLamar to have to modify documentation more than once for the same method, which is why I suggested this method documentation guide before working on more documentation improvements.

I'll use this PR as a starting point and try to submit my own PR for a method documentation guide by the end of the week.

**Comment by marcandre** [comment]
@ioquatix I don't have expectations about anybody contributing, but for those that do want to contribute, I certainly do expect your point 3 above (to address feedback).

Speaking for myself, I volunteer to improve Ruby and OSS primarily for enjoyment. I have done so for over a decade and I have contributed to many OSS projects. I try to communicate in a neutral tone (sometimes to mixed success). 
I've had many interactions with many people I didn't know, many very knowledgeable and very experienced and sometimes with criticism over things that were not clear cut. Yet I have difficulty recalling such resistance and feeling as frustrated as with the last discussions I've had with @BurdetteLamar. I also wish I was better at communicating and I hope that further conversations will be more satisfying for everyone involved.

**Comment by BurdetteLamar** [comment]
For my part, I don't think any offense has been intended, and certainly none taken.

All I want, for now, is to complete the last 15 of the 200 methods (in ENV, Hash, and Array) in the same style to which no one objected for eight months.

After that, it will be my pleasure to revisit any of this work when we have an agreed-upon guide in hand.

It's true, @ioquatix, that I do this work for the enjoyment.  But there's more.

[I'm going to digress.  But if one can't digress in a closed PR, then where?]

I worked in the software business for 40 years.  Some of the work was good and important -- at the time.  Now, nothing remains. Nothing I built or wrote is in current use.

These Ruby doc pages will remain relevant for a long time -- probably past my time (I'm 77).  So I like to think of this work as a legacy.

**Comment by jeremyevans** [comment]
> All I want, for now, is to complete the last 15 of the 200 methods (in ENV, Hash, and Array) in the same style to which no one objected for eight months.
> 
> After that, it will be my pleasure to revisit any of this work when we have an agreed-upon guide in hand.

Considering that, I think it's fine for you to finish documenting the remaining 15 methods.

**Pull request closed without merging**